### PR TITLE
iOS sdk documentation improvement

### PR DESCRIPTION
**Pull request by David-Henner**
It's really important for our partner to be able to integrate our sdk without our ios team support. Which is why I tried to integrate the iOS sdk in a new app with the help of the documentation only.
The aim was to be able to get a user tracked.

I updated the main README with Swift code examples and added a little extra info.
However, I didn't add Swift examples to the Error Handling section.

Overall the documentation was enough for me to get started on a really basic app (sdk configuration, signing up, logging in and starting the tracking).
I think it would be helpful to have a more complete documentation which would at least provide a general overview of the sdk's functionalities but could eventually explain them as well.

### CODE REVIEW

**Comment by conichi-ci-public** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">No CHANGELOG changes made</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">README.md modified</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: README.md [hunk 1]
```diff
@@ -67,17 +67,19 @@ Then, you must add a NSLocationAlwaysUsageDescription or NSLocationWhenInUseUsag
...
 
-And you're off! Take a look at the public documentation and start building. 
+And you're off! Take a look at the public documentation and start building.
 
 #### Initialization
 
-Example below shows the simplest setup `conichiSDK CoreKit` that points our `sandbox` server.
+Example below shows the simplest setup for `conichiSDK CoreKit` that points to our `sandbox` server.
+
+###### Objective-C
```
Comment (Antondomashnev) [inline README.md]: Would be great if we have an expanded sections for the code snippets like it's implemented here: https://github.com/SwiftGen/SwiftGen#installation
@David-Henner what do you think?

File: README.md [hunk 2]
```diff
@@ -96,11 +98,31 @@ Example below shows the simplest setup `conichiSDK CoreKit` that points our `san
   [CNISDK startWithConfiguration:configuration delegate:delegate];
 ```
 
-After this setup you have an access to the instance of the `[CNISDK sharedInstance]`.
+###### Swift
+
+```swift
+  //Create configuration for the sdk instance
+  let config: CNISDKConfiguration = CNISDKConfiguration.init(block: {
```
Comment (Antondomashnev) [inline README.md]: I think we can make it nicer with this code:
`let config: CNISDKConfiguration = CNISDKConfiguration.init() {` and without parameter. This is a usual way in Swift if you have clojure as a last function parameter.

File: README.md [hunk 3]
```diff
@@ -137,6 +161,47 @@ To start any real actions with guest `sdk` requires the guest to be authorized.
   }];
 ```
 
+###### Swift
+
+```swift
+// Create a signup request
+let info = CNISDKSignUpRequestInfo.init()
```
Comment (Antondomashnev) [inline README.md]: Do we need `init()` here? I think we can do just `CNISDKSignUpRequestInfo()`

File: README.md [hunk 4]
```diff
@@ -137,6 +161,47 @@ To start any real actions with guest `sdk` requires the guest to be authorized.
...
+info.firstName = "Jenessa"
+info.lastName = "Gretta"
+info.email = "jenessa.gretta@gmail.com"
+info.password = "strongestpasswordever=)"
+
+// Instantiate APImanager
+let apiManager = CNISDKAPIManager.init()
```
Comment (Antondomashnev) [inline README.md]: The same request as above 😄

File: README.md [hunk 5]
```diff
@@ -137,6 +161,47 @@ To start any real actions with guest `sdk` requires the guest to be authorized.
...
+apiManager.signUp(withRequest: info, completion: {
+  (guest, error)  in
+  if (error != nil) {
+    //handle error during the sign up
+  }
+  else {
+    //handle authorized guest
+  }          
+})
+```
```
Comment (Antondomashnev) [inline README.md]: I think we can make a last clojure parameter outside of the braces and also make optional unwrapping for error, so we have error object which is not optional inside if:
```swift
apiManager.signUp(withRequest: info) { (guest, error)  in
  if let error = error {
    //handle error during the sign up
  }
  else {
    handle authorized guest
  }          
}
```

File: README.md [hunk 6]
```diff
@@ -137,6 +161,47 @@ To start any real actions with guest `sdk` requires the guest to be authorized.
...
+
+`CNISDK.sharedInstance().startMonitoring()`
+
+###### Objective-C
+
+`[[CNISDK sharedInstance] startMonitoring];`
+
+Now, if a guest is near a beacon he should be tracked.
+
+customization of the tracking behavior is possible with callbacks. All callbacks can be found in `CNISDKDelegate.h`
```
Comment (Antondomashnev) [inline README.md]: Capital letter and maybe we don't need line space here 😄

File: README.md (hunk 1 above)
Comment (David-Henner) [inline README.md]: Great idea, it would bring more clarity to the document

File: README.md (hunk 2 above)
Comment (David-Henner) [inline README.md]: Yeah it makes sense, even better like that `let config: CNISDKConfiguration = CNISDKConfiguration() {`

File: README.md (hunk 3 above)
Comment (David-Henner) [inline README.md]: You're right we don't :)

File: README.md (hunk 5 above)
Comment (David-Henner) [inline README.md]: I see. Is there any other reasons to do it like that besides avoiding the risk of the optional being nil by the time we try to access its value ?

Comment (wladyslaws) [inline README.md]: yes, it helps you to follow "early return" rule, so developer won't be bothered by too much else-if clauses which are difficult to read. With optional unwrapping the error code may look like this:

```
if let error = error {
    //handle error during the sign up
} //remove "else"
//handle unauthorized guest

File: README.md [hunk 7]
```diff
@@ -51,7 +51,7 @@ Run `carthage update`, and you should now have the latest version of `CNISDKCore
 
 #### Partial Installation
```
Comment (wladyslaws) [inline README.md]: we can also move make expandable sections from installation sections, or it's an overkill?

File: README.md [hunk 8]
```diff
@@ -51,7 +51,7 @@ Run `carthage update`, and you should now have the latest version of `CNISDKCore
...
 
-Following the links below you can find the instructions in the __Getting Started__ section how to install specific `kit`:
+Following the links below you can find the instructions in the __Getting Started__ section on how to install a specific `kit`:
```
Comment (wladyslaws) [inline README.md]: I feel like information on kits installation should be preceded somehow by instruction on what "kit" is actually and why user may need it. Or a link to section below which will explain each kit separately.

File: README.md [hunk 9]
```diff
@@ -67,17 +67,19 @@ Then, you must add a NSLocationAlwaysUsageDescription or NSLocationWhenInUseUsag
```
Comment (wladyslaws) [inline README.md]: I know this is not the part of the PR but we should drop `NSLocationWhenInUseUsageDescription` from line 61 of this readme and instead warn user that app may not work properly if application is in "While using the app" mode of Location privacy settings. These were recent changes we had in our Guest app, and from production field collected feedback it appears SDK is not working as expected when having this setting enabled. @Mourad-Aly what do you think? @David-Henner can you update this as well? we don't work on readme much, and it's a good place to apply it painlessly.

File: README.md [hunk 10]
```diff
@@ -96,11 +98,31 @@ Example below shows the simplest setup `conichiSDK CoreKit` that points our `san
...
+
+After this setup you have an access to the instance of the `[CNISDK sharedInstance]` in **Obj-C** or  `CNISDK.sharedInstance()` in **Swift**.
 
 #### Error handling
 
-Conichi provides the unify error codes for each error comes from `sdk`. All error codes can be found in `CNIError.h` class.
+Conichi provides unified error codes for each error coming from the `sdk`. All error codes can be found in the `CNIError.h` class.
 To make `sdk` able to convert different error codes into `CNIErrorCode` there is a protocol `CNIErrorCodeConverter`. Out of the box there are 4 predefined converters:
 * CNINSURLErrorConverter
 * CNIHTTPResponseErrorConverter
```
Comment (wladyslaws) [inline README.md]: you don't have to work on error handling section, I will update it with new major version of the SDK.

File: README.md (hunk 7 above)
Comment (David-Henner) [inline README.md]: Yes I think it'd be nice to have expandable sections here as well

File: README.md (hunk 8 above)
Comment (David-Henner) [inline README.md]: what about nested expandable sections which would contain explanations about each kit ?

File: README.md (hunk 9 above)
Comment (David-Henner) [inline README.md]: so @wladyslaws should I just drop this part of the readme ?

File: README.md (hunk 4 above)
Comment (David-Henner) [inline README.md]: fixed here 7cc1b3c

File: README.md (hunk 6 above)
Comment (David-Henner) [inline README.md]: done here 341c969

File: README.md (hunk 5 above)
Comment (David-Henner) [inline README.md]: done here 341c969

File: README.md (hunk 4 above)
Comment (David-Henner) [inline README.md]: done here 341c969

File: README.md (hunk 3 above)
Comment (David-Henner) [inline README.md]: done here 341c969

File: README.md (hunk 2 above)
Comment (David-Henner) [inline README.md]: done here 341c969

File: README.md (hunk 9 above)
Comment (Mourad-Aly) [inline README.md]: We need to discuss together about the impact, i've created this https://conichi.atlassian.net/browse/CNI-5730 for @David-Henner for the 17's week

File: README.md [hunk 11]
```diff
@@ -136,6 +171,49 @@ To start any real actions with guest `sdk` requires the guest to be authorized.
...
+// Instantiate APImanager
+let apiManager = CNISDKAPIManager()
+
+// Perform signup
+apiManager.signUp(withRequest: info) {
+  (guest, error) in
+  if let error = error {
+    //handle error during the sign up
+  }
+  //handle authorized guest
```
Comment (Mourad-Aly) [inline README.md]: Do we need an `else` here? because now if there is an error, you're going to handle the error and also handle the authorized guest.

Comment (David-Henner) [inline README.md]: @Mourad-Aly either an `else` or a `return` in the `if` above. I forgot to add the `return` in my last changes. what do you think is better between the two ?

Comment (Mourad-Aly) [inline README.md]: @David-Henner Both are fine, we follow early `return` 👍

File: README.md (hunk 9 above)
Comment (wladyslaws) [inline README.md]: remove NSLocationWhenInUsageDescription as mandatory and inform that settings **cannot** include it for app to work properly.

**Comment by wladyslaws** [comment]
@David-Henner I suppose we are very close to merging it:) please address my comments and we go!

**Comment by Mourad-Aly** [comment]
retest this please

File: README.md (hunk 9 above)
Comment (David-Henner) [inline README.md]: @wladyslaws done here : fd4c52c

**Comment by Mourad-Aly** [comment]
@wladyslaws for me it's fine, do you have something required to do?

**Comment by wladyslaws** [comment]
good to go!

**Pull request merged**